File: 57-insert-interval/insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function(intervals, newInterval) {
    if (intervals.length === 0) return [newInterval]
    
    let inserted = false
    const merged = []

    for (let i = 0; i < intervals.length; i++) {
        const current = intervals[i]

        if (current[1] < newInterval[0]) {
            merged.push(current)
        } else if (current[0] > newInterval[1]) {
            if (!inserted) {
                merged.push(newInterval)
                inserted = true
            }
            merged.push(current)
        } else {
            newInterval = [
                Math.min(current[0], newInterval[0]),
                Math.max(current[1], newInterval[1])
            ]
        }
    }

    if (!inserted) {
        merged.push(newInterval)
    }

    return merged
};

File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        result = []
        i = 0
        n = len(intervals)

        while i < n and intervals[i][1] < newInterval[0]:
            result.append(intervals[i])
            i += 1

        merged_interval = list(newInterval)
        while i < n and intervals[i][0] <= merged_interval[1]:
            merged_interval[0] = min(merged_interval[0], intervals[i][0])
            merged_interval[1] = max(merged_interval[1], intervals[i][1])
            i += 1

        result.append(merged_interval)

        while i < n:
            result.append(intervals[i])
            i += 1

        return result